lldb-eval/eval_test.cc

```diff
@@ -653,6 +653,23 @@ TEST_F(EvalTest, TestPointerDereference) {
   EXPECT_THAT(Eval("*p_int0 + 1"), IsEqual("1"));
   EXPECT_THAT(Eval("*cp_int5"), IsEqual("5"));
   EXPECT_THAT(Eval("*cp_int5 - 1"), IsEqual("4"));
+
+  EXPECT_THAT(Eval("&*p_null"), IsEqual("0x0000000000000000"));
+  EXPECT_THAT(Eval("&p_null[4]"), IsEqual("0x0000000000000010"));
+  EXPECT_THAT(Eval("&*(int*)0"), IsEqual("0x0000000000000000"));
+  EXPECT_THAT(Eval("&((int*)0)[1]"), IsEqual("0x0000000000000004"));
+
+  EXPECT_THAT(Eval("**pp_int0"), IsEqual("0"));
+  EXPECT_THAT(Eval("**pp_int0 + 1"), IsEqual("1"));
+  EXPECT_THAT(Eval("&**pp_int0"), IsOk());
+  EXPECT_THAT(Eval("&**pp_int0 + 1"), IsOk());
+
+  EXPECT_THAT(Eval("&(true ? *p_null : *p_null)"),
+              IsEqual("0x0000000000000000"));
+  EXPECT_THAT(Eval("&(false ? *p_null : *p_null)"),
+              IsEqual("0x0000000000000000"));
+  EXPECT_THAT(Eval("&*(true ? p_null : nullptr)"),
```

Comment:
So what happens when you do &&**nullptr?

Reply:
Is not that bad? I thought &* pairs should cancel out.

I was wondering if it would not be more understandable if the evaluation methods returned pair of dereference-count and value, then you would get the actual value by applying dereference dereference-count-times on the value.

Address-of would be then calculated as simply decreasing the dereference count. If you needed the real value (e.g., to do arithmetic), you would just apply the dereferences.

Reply:
`& &**p` is not valid in C++, "&*" pair should not be eliminated in this case.

At first I tried to implement by doing the dereference lazily, i.e. have `Value` to know about pending derefs and doing it only when the value is actually requested. But that get tricky in other cases. For example, `&(*foo_ptr)->bar` should resolve to the `foo_ptr + offset_of(bar)` :/